### PR TITLE
Migrate classes talking to mongo to py3

### LINKED ISSUE #871: [Py2Py3] Migrate classes talking to MongoDB
**Impact of the new feature**
Almost every unified module

**Is your feature request related to a problem? Please describe.**
There are several classes talking to MongoDB in utils.py, which are used in Unified modules. We should migrate those to `src/python/Services/Mongo` (?).

**Describe the solution you'd like**
Here is the list of classes:
- statusHistory
- StartStopInfo
- batchInfo
- campaignInfo
- moduleLock
- remainingDatasetInfo
- reportInfo
- cacheInfo
- closeoutInfo
- wtcInfo

It seems that those classes share some methods, so I thought of creating a parent class to avoid duplicated code. 

**Describe alternatives you've considered**
None.

**Additional context**
@haozturk

### PR DESCRIPTION
**Pull request by amanda-grp**
Fixes #871 

#### Status
In development.

#### Description
Migration  to python3 of helper functions/classes using mongo in utils.py:
- `mongo_client` to Databases/Mongo/MongoClient.py
- `CampaignInfo` to Components/CampaignsConfigurationController.py
- `BatchInfo` to Components/BatchesController.py
- `StartStopInfo` to Components/RuntimeMonitor.py
- `ModuleLock` to Components//ModuleLock.py
- `RemainingDatasetInfo` to Components/RemainingDatasetsMonitor.py
- `CloseoutInfo` to Components/Workflow/WorkflowsCloseoutMonitor.py
- `ReportInfo` to Components/Workflow/WorkflowsReporter.py
- `StatusHistory` to Components/Workflow/WorkflowsStatusMonitor.py
- `WtcInfo` to Components/Workflow/WorkflowsTrafficController.py
- `AgentInfo` to Components/AgentsStatusController.py
- `CacheInfo` to Cache/CacheManager.py

This PR also includes the migration of:
- `utils.py/dataCache` to Cache/DataCacheLoader.py
- `utils.py/docCache` to Cache/DocumentCache.py
- `utils.py/eosFile` to Services/EOS/EOSWriter.py
- `utils.py/eosRead` to Services/EOS/EOSReader.py
- `utils.py/getAllAgents` to Services/WMStats/WMStatsReader.py
- `utils.py/sendAgentConfig` to Services/ReqMgr/ReqMgrWriter.py/ReqMgrWriter.setAgentConfig
- `utils.py/display_N` to Utilities/Logging.py/displayNumber
-  `display_time` to Utilities/Logging.py/displayTime
- `TrelloClient` to Services/Trello/TrelloClient.py
- `Unified/assignSession.py` to Databases/Oracle/OracleClient.py
- `Unified/assignSchema.py` to Databases/Oracle/OracleDB.py

#### To do:
- [ ] It's passing all unittest, but unit tests should be improved by using permanent data.
- [ ] Improve code structure
- [x] The AgentInfo depends on doc cache, so that the doc cache migration structure was included in this PR. This requires further verification on the doc keys as discussed in #883 and a review of the complete doc cache migrated in #899.
- [x] The CampaignInfo depends on the implementation of SiteInfo().allSite. This should be migrated in a future PR (#899)

#### Is it backward compatible (if not, which system it affects?)
No.

#### Related PRs
- The CacheManager implementation is a follow up of PR #867, which uses this module
- The WtcInfo and the AgentInfo depends on reqmgr helper functions migrated to python3 in PR #859 

#### Mention people to look at PRs
@haozturk

### CODE REVIEW

File: src/python/Services/WMStats/WMStatsReader.py [hunk 1]
```diff
@@ -4,14 +4,63 @@
...
     """
 
-    def __init__(self, url, logger=None, **contact):
-        pass
+    def __init__(self, logger: Optional[Logger] = None, **contact):
+        try:
+            configurationHandler = ConfigurationHandler()
+            self.reqmgrUrl = os.getenv("REQMGR_URL", configurationHandler.get("reqmgr_url"))
+            self.wmstatsEndpoint = {
+                "agentInfo": "/couchdb/wmstats/_design/WMStats/_view/agentInfo?stale=update_after"
```
Comment (amanda-grp) [inline src/python/Services/WMStats/WMStatsReader.py:29]: Hi, Alan! I'd like to ask you a question, is there any other endpoint that can replace this call against `couchdb`?

@amaltaro

Comment (amaltaro) [reply src/python/Services/WMStats/WMStatsReader.py:29]: Hi, Amanda. Nope, there is no alternative for this one, so we need to keep calling couchdb directly.

File: src/python/Cache/DocumentCache.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+class GWMSMONTotalsDocCache(DocumentCacheBuilder):
+    """
+    __GWMSMONTotalsDocCache__
+    General API for building the chaching document of key gwmsmon_totals
+    """
+
+    # TODO: url below does not return a json
+    def get(self) -> dict:
+        with os.popen("curl --retry 5 -s https://cms-gwmsmon.cern.ch/poolview/json/totals") as file:
```
Comment (amanda-grp) [inline src/python/Cache/DocumentCache.py:56]: Hasan, it is in this file that gwmsmon is being used and in the WorkflowInfo ("https://cms-gwmsmon.cern.ch/prodview/json/{wf}/summary"). I included the usage list in here: #883.

@haozturk

File: src/python/Operations/Trello/TrelloClient.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
+import os
+import json
+import logging
+from logging import Logger
+
+from typing import Optional
+
+
+class TrelloClient(object):
```
Comment (haozturk) [inline src/python/Operations/Trello/TrelloClient.py:9]: @amanda-grp I think it's better to place Trello under `Services` package and remove `Operations`. I imagine `Services` as the package which contains all the modules interacting w/ external services. Having an `Operations` package just complicates things and makes it harder to understand. What do you think? 

Having a `Databases` package makes sense to some extent , though. Databases are also external services which we use, but they are special and it makes sense to me to place them in a separate package.

Having a `Storage` package and placing `EOS` there is something that I'm not very sure of.  EOS is also a "Service" after all. Why not just place it `Services`?  What do you think about these organization in general? Does it make sense to you?

Comment (amanda-grp) [reply src/python/Operations/Trello/TrelloClient.py:9]: I think I agree with you about these points — although it still seems reasonable to me to place Trello and JIRA (not migrated yet)  together in a separate package for operational services.

Also, I'm not sure if everything in the `Components` package makes sense together, what do you think?

Comment (amanda-grp) [reply src/python/Operations/Trello/TrelloClient.py]: @haozturk EOS and Trello are in the Services package now

**Comment by haozturk** [comment]
@amanda-grp  Here is my proposal for the organization of `Components`:

1. `Components` should be renamed as `MongoControllers` since each module inside is responsible for controlling a mongodb collection. Components is vague and MongoControllers is pretty explicit
2. Rename the modules as following:

- `AgentsStatusController` --> `AgentController`: Let's open up a GH issue to revisit this module
- `BatchesController.py` --> `BatchController`
- `CampaignsConfigurationController` --> `CampaignController`
- `ModuleLock` --> `ModuleLockController`: Let's open up a GH issue to revisit this module
- `RemainingDatasetsMonitor` --> `RemainingDatasetController`:  Let's open up a GH issue to revisit this module
- `RuntimeMonitor` --> `StartStopTimeController`:  Let's open up a GH issue to revisit this module

3. Let's get rid of `Workflow` package and put those modules directly under `MongoControllers` and rename them as:

- `WorkflowsCloseoutMonitor.py` --> `CloseoutController`: Let's open up a GH issue to revisit this module
- `WorkflowsReporter.py` --> `ReportController`: Let's open up a GH issue to revisit this module
- `WorkflowsStatusMonitor.py` --> `StatusHistoryController`: Let's open up a GH issue to revisit this module
- `WorkflowsTrafficController.py` --> `WTCController`: Let's open up a GH issue to revisit this module

This organization makes way more sense to me. All these modules are doing more than monitoring, so `controller` is a better name. I suggested opening up GH issues for most of them, since these mongo collections and their controllers should be revisited later, it's not in the scope of this PR. If you agree and make the changes, I'll give a last round of review.

**Comment by amanda-grp** [comment]
Great, thank you very much @haozturk ! I'll update the notes

**Pull request merged**